psutil/_psosx.py

```diff
@@ -212,6 +213,84 @@ def disk_partitions(all=False):
 # =====================================================================
 
 
+def get_smc_temperatures(smc_keys, iter_num, iteratable):
+    """ Receives a dictionary of SMC keys, and a number possible keys
+    If keys can be iterated, replaces the expected %d in the key.
+    Returns a list of all smc_get_temperature results higher than 0
+    """
+    ret = list()
+    if iteratable:
+        for idx in range(iter_num):
+            for key, label in smc_keys.items():
+                result = cext.smc_get_temperatrue(key % idx)
+                if int(result) <= 0:
+                    continue
+                ret.append((label % idx, result, None, None))
+    else:
+        for key, label in smc_keys.items():
+            result = cext.smc_get_temperatrue(key)
+            if int(result) <= 0:
+                continue
+            ret.append((label, result, None, None))
+
+    return ret
+
+
+def sensors_temperatures():
+    """Return CPU temperatures
+    including key name and temperature.
+    """
+    ret = collections.defaultdict(list)
+    # Handle CPU cores
+    num_cpus = cext.cpu_count_phys()
+    core_temps = {
+        "TC%dC": "Core%d",
+    }
+    if num_cpus > 0:
+        results = get_smc_temperatures(core_temps, num_cpus + 1, True)
+        ret["CPU"] = ret["CPU"] + results
+
+    # Handle other CPU sensors
+    cpu_temps = {
+        "TC%dF": "CPU TC%dF",
+        "TC%dG": "CPU TC%dG",
+        "TC%dP": "CPU %d Proximity",
+    }
+    num_packages = cext.cpu_count_packages()
+    if num_packages > 0:
+        results = get_smc_temperatures(cpu_temps, num_packages, True)
+        ret["CPU"] = ret["CPU"] + results
+
+    # Handle Battery Temperature
+    battery_temps = {
+        "TB%dT": "Battery%d",
+    }
+    # 2 is the highest number of batteries recoreded in iStats
+    results = get_smc_temperatures(battery_temps, 3, True)
+    ret["Battery"] = ret["Battery"] + results
+
+    # Handle HDD temperatures
+    hdd_temps = {
+        "TH%dP": "HDD %d Proximity",
+        "TH%da": "HDD %d Unknown",
+    }
+    # 2 is a logically normal number of HDDs
+    results = get_smc_temperatures(hdd_temps, 3, True)
```

Comment:
mmm I suppose this is arbitrary; ...and I suppose we have no way to know the number of HDDs upfront as we do for fans, right?

Reply:
A short research shows you can call `diskutil list`, which on my system with a single drive returns:
```
/dev/disk0 (internal, physical):            
   #:                       TYPE NAME                    SIZE       IDENTIFIER           
   0:      GUID_partition_scheme                        *500.1 GB   disk0                
   1:                        EFI EFI                     209.7 MB   disk0s1              
   2:          Apple_CoreStorage Macintosh HD            499.2 GB   disk0s2              
   3:                 Apple_Boot Recovery HD             650.0 MB   disk0s3              

/dev/disk1 (internal, virtual):             
   #:                       TYPE NAME                    SIZE       IDENTIFIER           
   0:                  Apple_HFS Macintosh HD           +498.9 GB   disk1                
                                 Logical Volume on disk0s2                               
                                 795438B9-8D7D-42DC-92F4-4659049757DE                    
                                 Unencrypted                                             
```
So basically something like:  `diskutil list  | grep physical | wc -l` should work
Maybe there is a better way, or just leave it at 0 for now.

Reply:
Nope, it should be done in C. Forget about it for now if you can't figure out how to do it.